File: Java/CourseSchedule.java
public class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        if(prerequisites.length == 0)
            return true;
        int n = numCourses;
        int[][] graph = new int[n][n];
        
        for(int i=0; i<prerequisites.length; ++i) {
            int x = prerequisites[i][0];  // next class
            int y = prerequisites[i][1];  // prere class
            
            graph[y][x] = 1; // y -> x
        }
        
        return !detectCycle(graph);
    }

    public boolean detectCycle(int[][] graph){
        int n = graph.length;
        
        int[] topOrder = new int[n];
        int[] dfNum = new int[n];
        for(int i=0; i<n; ++i) {
            topOrder[i] = -1;
            dfNum[i] = -1;
        } 
        
        boolean[] cycle = {false};
        int[] currTopOrder = {n};
        int[] currDFNum = {0};
        
        for(int i = 0; i < n && !cycle[0]; ++i){
            DFS(graph, i, topOrder, dfNum, currTopOrder, currDFNum, cycle);
        }
        return cycle[0];
    }
    
    public void DFS(int[][] graph, int node, int[] topOrder, int[] dfNum, int[] currTopOrder, int[] currDFNum, boolean[] cycle){
        if(dfNum[node] != -1 || cycle[0]) {
            return;
        }
        
        ++currDFNum[0];
        dfNum[node] = currDFNum[0];
        
        for(int i = 0; i < graph.length && !cycle[0]; ++i) {
            if(graph[node][i] != 0 && dfNum[i] == -1) {
                DFS(graph, i, topOrder, dfNum, currTopOrder, currDFNum, cycle);
            } else if(graph[node][i] != 0 && dfNum[i] != -1 && topOrder[i] == -1){
                cycle[0] = true;
            }
        }
        
        topOrder[node] = currTopOrder[0];
        --currTopOrder[0];
    }
}

File: C++/course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int, vector<int>> G;
        
        for(auto c : prerequisites) G[c[0]].push_back(c[1]);
        
        vector<bool> haveTaken(numCourses);
        vector<bool> visited(numCourses);
        bool canTakeAll = true;
        for(int i=0; i < numCourses; ++i)
        {
            canTakeAll = canTakeAll && DFS(i, G, haveTaken, visited);
        }
        
        return canTakeAll;
    }
    
    bool DFS(int i, unordered_map<int, vector<int>> &G, vector<bool> &taken, vector<bool> &visited)
    {
        if(taken[i] || visited[i]) return taken[i];
        
        bool canTake = true;
        visited[i] = true;
        for(auto c : G[i])
        {
            canTake = canTake && DFS(c, G, taken, visited);
        }
        
        visited[i] = false;
        return taken[i] = canTake;
    }
       
};
